### DISCUSSION TITLE
How to attach `perf_events`?

### DISCUSSION THREAD

**Discussion by mrproliu** [cilium/ebpf#548] [Q&A] [answered]
In BCC, we have `BPF.attach_perf_event` to attach the `perf_event` and could use `sample_period` parameter to trigger with an interval.
But I can't see how to attach `perf_events` in cilium/ebpf?

**Comment by florianl** [comment, marked as answer]
`cilium/ebpf` is a Go implementation around the eBPF ecosystem and it is independent of the `perf` subsystem of the Linux kernel. BCC decided to do a mix and implement both but also decided to implement only a part of the complete `perf` userspace API.
Go implementations of the `perf` subsystem are for example [acln0/perf](github.com/acln0/perf) or [hodgesds/perf-utils](https://github.com/hodgesds/perf-utils) to name only two. Both packages can be used along with `cilium/ebpf` to attach eBPF programs to `perf` events.
Also the standard library provides an option to do so with [unix#PerfEventOpen](https://pkg.go.dev/golang.org/x/sys/unix#PerfEventOpen). 
General speaking there are usually four steps involved to attach an eBPF program to a perf event:

1. Open the perf event
2. Load the eBPF program
3. Set the eBPF program on the perf event
4. Enable the perf event

But there are some caveats to keep in mind. One can not attach an eBPF program to every perf event and attaching an eBPF program to a perf event represents only a very small subset of the `perf` subsystem.

**Reply by mrproliu** [reply]
Thanks for your professional response. 
I see. I have read the unix#PerfEventOpen, It could be used [unix.IoctlSetInt](https://pkg.go.dev/golang.org/x/sys/unix#IoctlSetInt) and use `unix#PERF_EVENT_IOC_SET_BPF` to set the BPF program to the event. 

> But there are some caveats to keep in mind. One can not attach an eBPF program to every perf event and attaching an eBPF program to a perf event represents only a very small subset of the perf subsystem.

In my case, I want to profile a `java` program, dump the User Space Stack It's enough. So I could use the `unix.PERF_COUNT_SW_CPU_CLOCK` as the event type, and use a fixed time as the sample/interval. Am I right?
But How can I link the BPF function to the perf event? Such as, If I want to monitor an `uprobe`, I could create a function and use `SEC("uprobe/xxx")` to make sure the function will be invoked. But If I use `perf_event`, Linux how to know which function will be used? 
Also, I want to have a mix monitor, such as when java `uprobe/JVM_InvokeMethod` then I record the thread id to a BPF map, I only need some thread dump data. Should I need to create another BPF program, and use the go program to mix the data?

**Reply by mrproliu** [reply]
Hi @florianl, I have tried to use the `unix.IoctlSetInt` to make it finished, but still got some problems. Here is some code(only some important code), the `eBPF` program is not invoked, If you want to see the full code, I have created [a new repo in Github](https://github.com/mrproliu/my_ebpf_test/tree/main/perf).

eBPF code:
```c++
SEC("perf_event")
int do_perf_event(struct pt_regs *ctx) {
    bpf_printk("hello");
    return 0;
}
```

Golang frontend:
```go
eventAttr := &unix.PerfEventAttr{
	Type:        unix.PERF_TYPE_SOFTWARE,
	Config:      unix.PERF_COUNT_SW_CPU_CLOCK,
	Sample_type: unix.PERF_SAMPLE_RAW,
	Sample:      1,
	Wakeup:      1,
}
fd, err := unix.PerfEventOpen(
	eventAttr,
	pid,
	0,
	-1,
	0,
)

objs := bpfObjects{}
if err := loadBpfObjects(&objs, nil); err != nil {
	log.Fatalf("loading objects: %s", err)
}
defer objs.Close()

// attach ebpf to perf event
unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_SET_BPF, objs.DoPerfEvent.FD())

// start perf event
unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_ENABLE, 0)

```

**Reply by florianl** [reply]
Here is a full working example for attaching an [eBPF program on perf event in Go](https://gist.github.com/florianl/5d9cc9dbb3822e03f6f65a073ffbedbb). It is structured very similar to the [tracepoint_in_go](https://github.com/cilium/ebpf/blob/master/examples/tracepoint_in_go/main.go) example from this repository except for ther perf event part.

**Reply by mrproliu** [reply]
I have tried your example code, but It still doesn't have any output. Is anything wrong?

![image](https://user-images.githubusercontent.com/3417650/150903903-ba48023e-8c95-4ad8-aa76-57c02657337f.png)

Is that mean my Linux system is not supported?
```shell
$ uname -a
Linux bpf-test1 4.19.0-18-cloud-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux
```

**Reply by ti-mo** [reply]
@mrproliu have you tried running this on a more recent kernel? There's a known issue with uprobes on older kernels: https://github.com/cilium/ebpf/issues/406. You're not setting uprobes here, but could be related, who knows. :shrug:

**Reply by mrproliu** [reply]
> @mrproliu have you tried running this on a more recent kernel? There's a known issue with uprobes on older kernels: #406. You're not setting uprobes here, but could be related, who knows. 🤷

I have printed the `uname -a` result, you could see it's >= `4.19`, do I miss anything?

**Reply by ti-mo** [reply]
I see you're using 4.19, hence my suggestion of trying something more recent.

**Reply by mrproliu** [reply]
I have updated the kernel, but it still not working, also don't have any output :(
```
$ uname -a
Linux bpf-test2 5.11.0-1028-gcp #32~20.04.1-Ubuntu SMP Wed Jan 12 20:08:27 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

**Reply by ti-mo** [reply]
I've marked @florianl's response as the answer, I think it nicely outlines how to implement what was asked in the original question.

The resulting thread has turned into troubleshooting, so not sure if we can continue discussing this here.
